### ISSUE TITLE
[bug] Global shortcuts are never triggered on Linux

### ISSUE THREAD

**Issue by AlexMikhalev** [tauri-apps/tao#307]
### Describe the bug

Even using API example it's possible to register global shortcut, but it never triggers on linux 
[This line](https://github.com/tauri-apps/tauri/blob/46f2eae8aad7c6a228eaf48480d5603dae6454b4/examples/api/src/components/Shortcuts.svelte#L16) never reached. 

### Reproduction

Compile latest tauri (git commit 46f2eae8aad7c6a228eaf48480d5603dae6454b4) and run api example as per readme. Observe lack of global shortcut trigger - both JS and Rust examples don't work.

### Expected behavior

Shortcut triggered - from JS and Rust

### Platform and versions

```shell
Operating System - Pop!_OS, version 20.04 X64
```

**Comment by AlexMikhalev** [comment]
Hurrah! I am going to test tonight.

**Comment by AlexMikhalev** [comment]
I think that broke now register in JS, if I run example/api and try to register global shortcut I get an empty array in Tauri console (screenshot) and error in JS console: 

```
[Error] Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'window.rpc.notify')
	promiseEmptyOnRejected (chunk-KSXZ7OFV.js:136)
	promiseReactionJob
```
![Screenshot from 2022-02-10 13-01-56](https://user-images.githubusercontent.com/50685/153413685-1ab44f83-a6b9-4a74-95f3-06dff0dfd68a.png)

```
api git:(next) ✗ cargo tree | grep tao
│       ├── tao v0.6.1
➜  api git:(next) ✗ 
```

**Comment by FabianLars** [comment]
@AlexMikhalev i think you need to re-build the api package (`cd tooling/api && yarn && yarn build`). This is unrelated to this tao change tho, and caused by the huge audit changes.

**Comment by AlexMikhalev** [comment]
I rebuild before reporting the bug. I also rebuild it following the advice above, but global shortcuts are not triggered or registering now.
 Promise rejection is at the start of the Tauri app, that's not related to the bug above.

**Comment by amrbashir** [comment]
@AlexMikhalev the promise rejection error is as stated by Fabian because of outdated api package. Please wait for the new release which should be so soon or test the globalShortcut functionality using tao directly, you can find an example in the `examples` directory in tao repo.

**Comment by AlexMikhalev** [comment]
@amrbashir ok, waiting for the new release of Tauri with bugfix. This probably means the bug isn't closed yet.

**Comment by FabianLars** [comment]
@AlexMikhalev We released the first rc version yesterday 🥳

**Comment by wusyong** [comment]
We are looking for help! I couldn't reproduce this in my PC, laptop and even VM. But there are still people have this issue.
On x11, we listen [XNextEvent](https://github.com/tauri-apps/tao/blob/dev/src/platform_impl/linux/global_shortcut.rs#L62) here. And according to people that has trouble, XPending is always 0 for some reason despite they already typed several different keys.
And on wayland, we simply haven't done it yet. Contribution is welcome!

**Comment by jplatte** [comment]
What does "global" shortcut mean? Wayland intentionally doesn't let you read keyboard events when your window is not focused.

**Comment by wusyong** [comment]
Yeah, you are right. I forgot to mention tauri-apps/global-hotkey#28 too. wayland-protocols has an [RFC](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/56) about this but doesn't seem pretty active recently.

**Comment by FabianLars** [comment]
For whatever reason github didn't save my message so i try my best to replicate it from memory 😅 
> Wayland intentionally doesn't let you read keyboard events

Yep we know that, but:
- global shortcut don't work for some users on x11 either (this is what this issue was about)
- the current implementation doesn't work for focused wayland windows either
- ~~**if i remember correctly**, we found a workaround to make _global_ shortcuts work in a wayland session via winit's DeviceEvent::Key~~

**Comment by amrbashir** [comment]
> if i remember correctly, we found a workaround to make global shortcuts work in a wayland session via winit's DeviceEvent::Key

That is not right, `DeviceEvent::Key` is not supported on wayland.

This issue should just focus on x11 and wait for wayland to implement a protocol for global device events.

**Comment by FabianLars** [comment]
Yeah you're right, just checked our convo again and I wasn't even able to spawn a winit window on Wayland back then (it still doesn't work and it's still crashing my system 😂)

_but_ the device events were working correctly on (x11) systems where our current shortcut implementation doesn't work, so that's what I mixed up in my previous comment.

**Comment by Beanow** [comment]
So, on Manjaro Gnome X11 neither `tauri/examples/api` nor `cargo run --example=global_shortcut` here seem to have working shortcuts.

That said, it's the first time I switched to X11 (over wayland) on this install. And it's had more shortcut issues as I've messed about with ibus/fcitx for IME's. I'll see if I can reproduce elsewhere.

Any particular aspects you want me to test?

**Comment by Beanow** [comment]
So I tested on a manjaro live image, and shortcuts do work under Gnome X11, probably did bork something on the install.

![image](https://user-images.githubusercontent.com/497556/161775410-869dcbbe-5de0-4de5-b3aa-ec7ac48f28fc.png)

Edit: I did try fcitx5 and even pipewire to be sure, like I use on the installed version. :shrug: just works. Even when focussed under wayland and global on X11.

**Comment by Beanow** [comment]
> On x11, we listen [XNextEvent](https://github.com/tauri-apps/tao/blob/dev/src/platform_impl/linux/global_shortcut.rs#L62) here. And according to people that has trouble, XPending is always 0 for some reason despite they already typed several different keys.

Like mentioned here, on the installed OS I also see this.
XPending is always 0, and if I were to skip the XPending > 0 check XNextEvent blocks indefinitely.

Also strange, the Tauri API example defaults to Ctrl+X (which is cut by default), and plays the 'drip' alert sound when pressed. Like it tried and failed to cut. Which it didn't do when the binding was working properly.

I've not yet been able to find what caused it to break.

**Comment by amrbashir** [comment]
@AlexMikhalev do you still have this issue? If so can you check if toggling `NumLock` on and off makes any difference? shortcuts should trigger when it is off but will fail when is on, see tauri-apps/tao#537

**Comment by FabianLars** [comment]
For what it's worth, i can confirm that turning off numlock makes it work again. Took me some time to figure out that even tho my new keyboard doesn't have a numpad, num lock was still set to `on` (`xset q | grep Caps`), had to disable it via `xdotool key Num_Lock`

**Comment by amrbashir** [comment]
tauri-apps/tao#538 should fix this. If you want to test the fix, add this to your `Cargo.toml` and then run `cargo update -p tao` in `src-tauri`

Edit: the fix is now part of the dev branch
```toml
[patch.crates-io]
tao = { git = "https://github.com/tauri-apps/tao", branch = "dev" }
```

**Comment by AlexMikhalev** [comment]
Let me test during the weekend I will confirm if it's fixed.

**Comment by amrbashir** [comment]
Looks like the issue still exists but it definitely has something to do with NumLock On/Off state

**Comment by amrbashir** [comment]
Looks like this is a known X11 behavior, https://stackoverflow.com/questions/4037230/global-hotkey-with-x11-xlib.

**Comment by amrbashir** [comment]
A new fix is up in tauri-apps/tao#540, please test and give feedback, I will not merge it this time until I get confirmation.

@AlexMikhalev @beyondlov1
```toml
[patch.crates-io]
tao = { git = "https://github.com/tauri-apps/tao", branch = "dev" }
```

Edit: fix is now part of `dev` branch

**Comment by beyondlov1** [comment]
Great! It works. Thanks!  @amrbashir 

PS: As the dependencies of tauri (1.0.5) published on crates.io is :  tauri (1.0.5) -> tauri-runtime-wry (^0.10.2) -> wry(^0.19) -> tao(^0.12),  another patch should be added:

```
[patch.crates-io]
tauri = { git = "https://github.com/tauri-apps/tauri", branch = "dev"}
tao = { git = "https://github.com/tauri-apps/tao", branch = "dev" }
```
